apm-agent-core/src/main/java/co/elastic/apm/agent/bci/bytebuddy/PatchBytecodeVersionTo51Transformer.java

```diff
@@ -48,12 +49,31 @@
  * {@link Advice.WithCustomMapping#bootstrap} which is important for {@linkplain TracerAwareInstrumentation#indyPlugin() indy plugins}.
  */
 public class PatchBytecodeVersionTo51Transformer implements AgentBuilder.Transformer {
+
+    private static boolean isAtLeastJava7(TypeDescription typeDescription) {
+        ClassFileVersion classFileVersion = typeDescription.getClassFileVersion();
+        return classFileVersion != null && classFileVersion.getJavaVersion() >= 7;
+    }
+
     @Override
     public DynamicType.Builder<?> transform(DynamicType.Builder<?> builder, TypeDescription typeDescription, ClassLoader classLoader, JavaModule javaModule) {
+        if (isAtLeastJava7(typeDescription)) {
+            // we can avoid the expensive (and somewhat dangerous) stack frame re-computation if stack frames are already
+            // present in the bytecode, which also allows eagerly loading types that might be present in the method
+            // body, but not yet loaded by the JVM.
+            return builder;
+        }
         return builder.visit(new AsmVisitorWrapper.AbstractBase() {
             @Override
             public ClassVisitor wrap(TypeDescription typeDescription, ClassVisitor classVisitor, Implementation.Context context,
                                      TypePool typePool, FieldList<FieldDescription.InDefinedShape> fieldList, MethodList<?> methodList, int writerFlags, int readerFlags) {
+
+                if (isAtLeastJava7(typeDescription)) {
+                    // in this case we won't prevent recomputing stack frames, but at least we avoid extra visitor wrapping
+                    // this happens for example with db2 JDBC driver
+                    return classVisitor;
+                }
+
                 return new ClassVisitor(Opcodes.ASM7, classVisitor) {
                     private boolean patchVersion;
```

Comment:
At this point we already know `version < Opcodes.V1_7`, so this flag and related checks are not required anymore in this `ClassVisitor`, right?

Reply:
Good point, I've removed this extra check.

One very minor case might be when the class file version is not yet known when the `transform` method is executed but is known when the `AsmVisitorWrapper.wrap` method is being executed. That would be very unlikely though.

Reply:
I didn't get that - we are talking about a check within `wrap()` and then checks within the `visit` methods. Do we have a reason to believe the version we get in `wrap()` can somehow be different from the one the `visit()` method gets?

My actual suggestion was to keep your check and remove the ones in the `visit` methods